server/sonar-server/src/main/java/org/sonar/server/projectanalysis/ws/DeleteAction.java

```diff
@@ -0,0 +1,97 @@
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+package org.sonar.server.projectanalysis.ws;
+
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.stream.Stream;
+import org.sonar.api.server.ws.Request;
+import org.sonar.api.server.ws.Response;
+import org.sonar.api.server.ws.WebService;
+import org.sonar.api.web.UserRole;
+import org.sonar.core.util.Uuids;
+import org.sonar.db.DbClient;
+import org.sonar.db.DbSession;
+import org.sonar.db.component.SnapshotDto;
+import org.sonar.server.exceptions.NotFoundException;
+import org.sonar.server.user.UserSession;
+import org.sonarqube.ws.client.projectanalysis.DeleteRequest;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.String.format;
+import static org.sonar.db.component.SnapshotDto.STATUS_UNPROCESSED;
+import static org.sonarqube.ws.client.projectanalysis.ProjectAnalysesWsParameters.PARAM_ANALYSIS;
+
+public class DeleteAction implements ProjectAnalysesWsAction {
+  private final DbClient dbClient;
+  private final UserSession userSession;
+
+  public DeleteAction(DbClient dbClient, UserSession userSession) {
+    this.dbClient = dbClient;
+    this.userSession = userSession;
+  }
+
+  @Override
+  public void define(WebService.NewController context) {
+    WebService.NewAction action = context.createAction("delete")
+      .setDescription("Delete an analysis.<br>" +
+        "Requires one of the following permissions:" +
+        "<ul>" +
+        "  <li>'Administer System'</li>" +
+        "  <li>'Administer' rights on the specified project</li>" +
+        "</ul>")
+      .setSince("6.3")
+      .setPost(true)
+      .setHandler(this);
+
+    action.createParam(PARAM_ANALYSIS)
+      .setExampleValue(Uuids.UUID_EXAMPLE_04)
+      .setRequired(true);
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    Stream.of(request)
+      .map(toWsRequest())
+      .forEach(deleteAnalysis());
+
+    response.noContent();
+
+  }
+
+  private static Function<Request, DeleteRequest> toWsRequest() {
+    return request -> new DeleteRequest(request.mandatoryParam(PARAM_ANALYSIS));
+  }
+
+  private Consumer<DeleteRequest> deleteAnalysis() {
+    return request -> {
+      try (DbSession dbSession = dbClient.openSession(false)) {
+        SnapshotDto analysis = dbClient.snapshotDao().selectByUuid(dbSession, request.getAnalysis())
+          .orElseThrow(() -> new NotFoundException(format("Analysis '%s' not found.", request.getAnalysis())));
+        userSession.checkComponentUuidPermission(UserRole.ADMIN, analysis.getComponentUuid());
+        checkArgument(!analysis.getLast(), "The last analysis '%s' cannot be deleted", request.getAnalysis());
```

Comment:
If the analysis has already been deleted but the purge has not been executed, no error is reported.
It's a corner case, but I'm wondering if we should not fail in such cases.